File: MAP/Laborator3/src/service/Consola.java
package service;

import domain.Student;
import domain.Tema;
import repository.ValidationException;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;

public class Consola {
    private Service serv=new Service("./src/data/Studenti.txt","./src/data/Teme.txt");
    private BufferedReader br=new BufferedReader(new InputStreamReader(System.in));

    private void meniu(){
        System.out.println("Meniu");
        System.out.println("1. Adauga student");
        System.out.println("2. Sterge student");
        System.out.println("3. Actualizare date student");
        System.out.println("4. Cauta student");
        System.out.println("5. Vizualizare lista studenti");
        System.out.println("6. Adauga tema");
        System.out.println("7. Prelungire deadline");
        System.out.println("0. Exit");
    }

    private void callAdaugaStudent(){
        try {
            System.out.print("ID: ");
            String id=br.readLine();
            System.out.print("Nume: ");
            String nume=br.readLine();
            System.out.print("Grupa: ");
            String grupa=br.readLine();
            System.out.print("Email: ");
            String email=br.readLine();
            System.out.print("Profesor indrumator: ");
            String prof=br.readLine();
            if(serv.adaugaStudent(id,nume,grupa,email,prof)!=null) System.out.println("Studentul nu a putut fi adaugat");
            else System.out.println("Studentul a fost adaugat");
        } catch (IOException e) {
            e.printStackTrace();
        }catch (ValidationException e){
            System.out.println(e.getMessage());
        }
    }

    private void callStergeStudent(){
        try {
            System.out.print("ID student: ");
            String id=br.readLine();
            if(serv.stergeStudent(id)==null) System.out.println("Nu exista niciun student cu acest id");
            else System.out.println("Studentul a fost sters");
        } catch (IOException e) {
            e.printStackTrace();
        }catch (ValidationException e){
            System.out.println(e.getMessage());
        }
    }

    private void callActualizareStudent(){
        try {
            System.out.print("ID student: ");
            String id=br.readLine();
            Student s=serv.cautaStudent(id);
            if(s==null) System.out.println("Nu exista niciun student cu acest id");
            else{
                boolean updated=false;

                System.out.print("Nume: ");
                String nume=br.readLine();

                System.out.print("Grupa: ");
                String grupa=br.readLine();

                System.out.print("Email: ");
                String email=br.readLine();

                System.out.print("Profesor indrumator: ");
                String prof=br.readLine();

                if(!serv.actualizareStudent(id,nume,grupa,email,prof)) System.out.println("Nu s-au introdus date noi");
                else System.out.println("Datele studentului au fost modificate");
            }
        } catch (IOException e) {
            e.printStackTrace();
        }catch (ValidationException e){
            System.out.println(e.getMessage());
        }
    }

    private void callCautaStudent(){
        try {
            System.out.print("ID student: ");
            String id=br.readLine();
            Student s=serv.cautaStudent(id);
            if(s==null) System.out.println("Nu exista niciun student cu acest numar matricol");
            else System.out.println("Studentul cautat este "+s);

        } catch (IOException e) {
            e.printStackTrace();
        }catch (ValidationException e){
            System.out.println(e.getMessage());
        }
    }

    private void callListaStudenti(){
        for (Student s: serv.listaStudenti() ) {
            System.out.println(s);
        }
    }


    private void callAdaugaTema() {
        try {
            System.out.print("ID: ");
            String id = br.readLine();
            System.out.print("Descriere: ");
            String descriere = br.readLine();
            System.out.print("Deadline: ");
            String deadline = br.readLine();
            System.out.print("Predare: ");
            String predare = br.readLine();
            if (serv.adaugaTema(id, descriere, deadline, predare) == null)
                System.out.println("Tema adaugata");
            else System.out.println("Tema existenta");
        } catch (IOException e) {
            e.printStackTrace();
        }
        catch(NumberFormatException e){
            System.out.println("Dati numere pentru id,deadline si data predarii");
        }
        catch (ValidationException e){
            System.out.println(e.getMessage());
        }

    }
    private void callPrelungireDeadline(){
        try {
            System.out.print("ID: ");
            String id=br.readLine();
            System.out.print("Deadline nou: ");
            String deadline=br.readLine();
            if(serv.prelungireDeadLine(id,deadline)==null) System.out.println("Problemee");
            else System.out.println("Deadline actualizat");

        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void run(){
        while(true){
            meniu();
            System.out.print("Comanda: ");
            String cmd;
            try {
                cmd=br.readLine();
            } catch (IOException e) {
                System.out.println("Comanda invalida!");
                continue;
            }
            if (cmd.equals("0")) {
                System.out.println("Bye!");
                break;
            }
            if (cmd.equals("1")) {
                callAdaugaStudent();
                continue;
            }
            if (cmd.equals("2")) {
                callStergeStudent();
                continue;
            }
            if (cmd.equals("3")) {
                callActualizareStudent();
                continue;
            }
            if (cmd.equals("4")) {
                callCautaStudent();
                continue;
            }
            if (cmd.equals("5")) {
                callListaStudenti();
                continue;
            }
            if (cmd.equals("6")){
                callAdaugaTema();
                continue;
            }
            if(cmd.equals("7")){
                callPrelungireDeadline();
                continue;
            }
            System.out.println("Comanda invalida!");
        }

    }

}

File: MAP/Laborator11-14/CatalogMAP/CatalogMAP/consola/Consola.cs
using CatalogMAP.domain;
using CatalogMAP.service;
using CatalogMAP.validator;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace CatalogMAP.consola
{
    class Consola
    {
        private Service service;

        public Consola(Service service)
        {
            this.service = service;
        }

        private void Meniu()
        {
            Console.WriteLine("----------- Meniu -----------");
            Console.WriteLine("1. Adauga student");
            Console.WriteLine("2. Sterge student");
            Console.WriteLine("3. Actualizare date student");
            Console.WriteLine("4. Cauta student");
            Console.WriteLine("5. Vizualizare lista studenti");
            Console.WriteLine("6. Adauga tema");
            Console.WriteLine("7. Prelungire deadline");
            Console.WriteLine("8. Vizualizare lista teme");
            Console.WriteLine("9. Asignare nota");
            Console.WriteLine("10. Filtrare dupa grupa");
            Console.WriteLine("11. Filtrare teme dupa interval");

            Console.WriteLine("12. Media studentilor la laborator");
            Console.WriteLine("13. Studentii care intra in examen");
            Console.WriteLine("14. Studentii care au predat toate temele la timp");
           // Console.WriteLine("15. Media per laborator");
            
            Console.WriteLine("0. Exit");
            Console.WriteLine("-----------------------------");
        }

        private void CallAdaugaStudent()
        {
            try
            {
                Console.Write("ID: ");
                string id = Console.ReadLine();
                Console.Write("Nume: ");
                string nume = Console.ReadLine();
                Console.Write("Grupa: ");
                string grupa = Console.ReadLine();
                Console.Write("Email: ");
                string email = Console.ReadLine();
                Console.Write("Profesor indrumator: ");
                string prof = Console.ReadLine();
                Student student = new Student()
                {
                    ID = id,
                    Nume = nume,
                    Grupa = grupa,
                    Email = email,
                    IndrumatorLab = prof
                };
                bool adaugat = service.AdaugaStudent(student);

                if (adaugat)
                    Console.WriteLine(">> Studentul a fost adaugat!");
                else
                    Console.WriteLine(">> Studentul nu a putut fi adaugat!");
            }
            catch (ValidationException e)
            {
                Console.WriteLine(">> " + e.Message);
            }
        }

        private void CallStergeStudent()
        {
            try
            {
                Console.Write("ID student: ");
                String id = Console.ReadLine();
                bool sters = service.StergeStudent(id);
                if (sters)
                    Console.WriteLine(">> Studentul a fost sters");
                else Console.WriteLine(">> Nu exista niciun student cu id-ul " + id);
            }
            catch (ValidationException e)
            {
                Console.WriteLine(">> " + e.Message);
            }
        }

        private void CallActualizareStudent()
        {
            try
            {
                Console.Write("ID student: ");
                String id = Console.ReadLine();
                Student student = service.CautaStudent(id);
                if (student == default(Student))
                    Console.Write("Nu exista niciun student cu acest id");
                else {
                    bool updated = false;

                    Console.Write("Nume: ");
                    String nume = Console.ReadLine();
                    if (nume == "") nume = student.Nume;
                    else updated = true;

                    Console.Write("Grupa: ");
                    String grupa = Console.ReadLine();
                    if (grupa == "") grupa = student.Grupa;
                    else updated = true;

                    Console.Write("Email: ");
                    String email = Console.ReadLine();
                    if (email == "") email = student.Email;
                    else updated = true;

                    Console.Write("Profesor indrumator: ");
                    String prof = Console.ReadLine();
                    if (prof == "") prof = student.IndrumatorLab;
                    else updated = true;

                    if (!updated) Console.WriteLine(">> Nu s-au introdus date noi");
                    else
                    {
                        Student studentNou = new Student
                        {
                            ID = id,
                            Nume = nume,
                            Grupa = grupa,
                            Email = email,
                            IndrumatorLab = prof
                        };
                        service.ActualizeazaStudent(studentNou);
                        Console.WriteLine(">> Datele studentului au fost modificate");
                    }
                }
            }
            catch (ValidationException e)
            {
                Console.WriteLine(">> " + e.Message);
            }
        }

        

        private void CallCautaStudent()
        {
            try
            {
                Console.Write("ID student: ");
                string id = Console.ReadLine();
                Student student = service.CautaStudent(id);
                if (student == default(Student))
                    Console.WriteLine(">> Nu exista niciun student cu numarul matricol " + id);
                else Console.WriteLine(">> Studentul cautat este " + student);

            }
            catch (ValidationException e)
            {
                Console.WriteLine(">> " + e.Message);
            }
        }

        private void CallListaStudenti()
        {
            Console.WriteLine(">> Lista de studenti: ");
            int nrc = 1;
            foreach (Student s in service.ListaStudenti())
            {
                Console.WriteLine(nrc.ToString() + ". " + s);
                nrc++;
            }
        }

        private void CallAdaugaTema()
        {
            Console.Write("ID: ");
            string id = Console.ReadLine();
            Console.Write("Descriere: ");
            string descriere = Console.ReadLine();
            Console.Write("Deadline: ");
            string deadline = Console.ReadLine();
            Console.Write("Predare: ");
            string predare = Console.ReadLine();
            Tema tema = new Tema()
            {
                ID = id,
                Descriere = descriere,
                Deadline = deadline,
                DataPredare = predare
            };
            try
            {
                if (service.AdaugaTema(tema))
                    Console.WriteLine(">> Tema adaugata");
                else Console.WriteLine(">> Tema existenta!");
            }
            catch (ValidationException e)
            {
                Console.WriteLine(">> " +e.Message);
            }

        }
        private void CallPrelungireDeadline()
        {
            Console.Write("ID: ");
            string id = Console.ReadLine();
            Console.Write("Deadline nou: ");
            string deadline = Console.ReadLine();
            if (service.PrelungireDeadline(id, deadline))
                Console.WriteLine("Deadline actualizat");
            else Console.WriteLine("Deadline-ul nu a fost modificat!");

        }

        private void CallListaTeme()
        {
            Console.WriteLine(">> Lista de teme: ");
            int nrc = 1;
            foreach (Tema t in service.ListaTeme())
            {
                Console.WriteLine(nrc.ToString() + ". " + t);
                nrc++;
            }
        }

        private void CallAsignareNota()
        {
            try
            {
                Console.Write("ID student: ");
                string idS = Console.ReadLine();
                Console.Write("ID tema: ");
                string idT = Console.ReadLine();
                Console.Write("Nota: ");
                string notaProf = Console.ReadLine();
                Console.Write("Data: ");
                string data = Console.ReadLine();

                bool motivat = false;
                Console.WriteLine("Studentul a lipsit motivat? (y/n)");
                string ans = Console.ReadLine();
                if (ans.Equals("y") || ans.Equals("Y"))
                    motivat = true;

                Nota nota = new Nota()
                {
                    StudentID = idS,
                    TemaID = idT,
                    DataCurenta = data,
                    NotaProf = notaProf
                };
                if (!service.AdaugaNota(nota, motivat))
                    Console.WriteLine(">> Nota nu a putut fi adaugata!");
                else
                {
                    if (motivat)
                    {
                        Console.WriteLine(">> Nota maxima " + nota);
                        Console.WriteLine(">> Studentul a primit nota " + nota);
                    }
                    else
                    {
                        Tema t = service.CautaTema(idT);
                        Console.WriteLine(">> Nota maxima " + service.CalculeazaNota(data, "10", t));
                        Console.WriteLine(">> Studentul a primit nota " + service.CalculeazaNota(data, notaProf, t));
                    }
                }
            }
            catch (ValidationException e)
            {
                Console.WriteLine(">> " + e.Message);
            }
        }
    
        private void CallFiltrareGrupa()
        {
            Console.Write("Grupa: ");
            string grupa = Console.ReadLine();
            foreach (Student student in service.FilterByGrupa(grupa))
            {
                Console.WriteLine(student.ID+" "+student.Nume+" "+student.Email);
                
            }
        }

        private void CallFiltrareInterval()
        {
            Console.Write("Start: ");
            string start = Console.ReadLine();
            Console.Write("End: ");
            string end = Console.ReadLine();
            foreach (Tema t in service.FilterByInterval(start, end)) 
            {
                Console.WriteLine(t);

            }
        }

        private void CallMedieStudenti()
        {
            List<KeyValuePair<Student, double>> lista = service.MediaStudentilor();
            foreach (KeyValuePair<Student, double> pair in lista)
            {
                Console.WriteLine(pair.Key.Nume + " " + pair.Value);
            }
        }

        private void CallExamen()
        {
            List<KeyValuePair<Student, double>> lista = service.ListaExamen();
            foreach (KeyValuePair<Student, double> pair in lista)
            {
                Console.WriteLine(pair.Key.Nume + " - media:" + pair.Value);
            }
        }

        private void CallStudentiLaTimp()
        {
            List<Student> lista = service.ListaStudentiTemeLaTimp();
            foreach (Student s in lista)
            {
                Console.WriteLine(s.Nume);
            }
        }


        public void Run()
        {
            while (true)
            {
                Meniu();
                Console.Write("Comanda: ");
                string cmd;
                cmd = Console.ReadLine();
                
                if (cmd.Equals("0"))
                {
                    Console.WriteLine("Bye!");
                    break;
                }
                if (cmd.Equals("1"))
                {
                    CallAdaugaStudent();
                    continue;
                }
                if (cmd.Equals("2"))
                {
                    CallStergeStudent();
                    continue;
                }
                if (cmd.Equals("3"))
                {
                    CallActualizareStudent();
                    continue;
                }
                if (cmd.Equals("4"))
                {
                    CallCautaStudent();
                    continue;
                }
                if (cmd.Equals("5"))
                {
                    CallListaStudenti();
                    continue;
                }
                if (cmd.Equals("6"))
                {
                    CallAdaugaTema();
                    continue;
                }
                if (cmd.Equals("7"))
                {
                    CallPrelungireDeadline();
                    continue;
                }
                if (cmd.Equals("8"))
                {
                    CallListaTeme();
                    continue;
                }
                if (cmd.Equals("9"))
                {
                    CallAsignareNota();
                    continue;
                }
                if (cmd.Equals("10"))
                {
                    CallFiltrareGrupa();
                    continue;
                }
                if (cmd.Equals("11"))
                {
                    CallFiltrareInterval();
                    continue;
                }
                if (cmd.Equals("12"))
                {
                    CallMedieStudenti();
                    continue;
                }
                if (cmd.Equals("13"))
                {
                    CallExamen();
                    continue;
                }
                if (cmd.Equals("14"))
                {
                    CallStudentiLaTimp();
                    continue;
                }
                if (cmd.Equals("15"))
                {
                    
                    continue;
                }
                Console.WriteLine("Comanda invalida!");
            }

        }
    
    }
}
